.circleci/config.yml

```diff
@@ -0,0 +1,305 @@
+version: 2
+
+references:
+  container_config: &container_config
+    docker:
+      - image: aetrnty/builder
+        user: builder
+    working_directory: ~/epoch
+
+  rebar_cache_key: &rebar_cache_key rebar-cache-{{ checksum "rebar.lock" }}-{{ checksum "rebar.config" }}
+  restore_rebar_cache: &restore_rebar_cache
+    restore_cache:
+      key: *rebar_cache_key
+
+  restore_build_cache: &restore_build_cache
+    restore_cache:
+      key: build-cache-v1
+
+  packages_workspace: &packages_workspace /tmp/packages
+  build_package: &build_package
+    run:
+      name: Build Package Tarball
+      environment:
+        PACKAGES_DIR: *packages_workspace
+      command: |
+        if [ "$(uname -s)" == "Darwin" ]; then
+            PKG_SUFFIX="osx-$(sw_vers -productVersion)"
+        elif [ "$(uname -s)" == "Linux" ]; then
+            PKG_SUFFIX="linux"
+        fi
+        make prod-package
+        mkdir ${PACKAGES_DIR:?}
+        cp _build/prod/rel/epoch/epoch-$(cat VERSION).tar.gz ${PACKAGES_DIR:?}/epoch-$(cat VERSION)-${PKG_SUFFIX}.tar.gz
+
+  test_package: &test_package
+    run:
+      name: Test Package Tarball
+      environment:
+        PACKAGES_DIR: *packages_workspace
+      command: |
+        if [ "$(uname -s)" == "Darwin" ]; then
+            PKG_SUFFIX="osx-$(sw_vers -productVersion)"
+        elif [ "$(uname -s)" == "Linux" ]; then
+            PKG_SUFFIX="linux"
+        fi
+        epmd -daemon
+        make python-env
+        make python-release-test TARBALL=${PACKAGES_DIR:?}/epoch-$(cat VERSION)-${PKG_SUFFIX}.tar.gz
+
+  store_package_artifacts: &store_package_artifacts
+    store_artifacts:
+      path: *packages_workspace
+      destination: /
+
+  deploy_steps: &deploy_steps
+    - checkout
+    - attach_workspace:
+        at: *packages_workspace
+    - run:
+        name: Install PIP Requirements
+        command: |
+          pip install --user -r deployment/ansible/pip-requirements.txt
+    - run:
+        name: Deploy
+        environment:
+          PACKAGES_DIR: *packages_workspace
+        command: |
+          export PACKAGE=${PACKAGES_DIR:?}/epoch-$(cat VERSION)-linux.tar.gz
+          cd deployment/ansible
+          ansible-playbook -i inventory/openstack.yml --limit="epoch:&${DEPLOY_ENV:?}" \
+            --extra-vars "package=${PACKAGE:?} env=${DEPLOY_ENV:?} \
+              tester_keys_password='${tester_keys_password:?}' datadog_api_key=${datadog_api_key:?} datadog_app_key=${datadog_app_key:?}" \
+            deploy.yml
+
+jobs:
+  build:
+    <<: *container_config
+    steps:
+      - checkout
+      - *restore_rebar_cache
+      - run:
+          name: Build
+          command: ./rebar3 as test do release
+      - save_cache:
+          key: build-cache-v1
+          paths:
+            - "_build"
+            - "apps/aecuckoo/c_src"
+            - "apps/aecuckoo/priv/bin"
+            - "apps/aecuckoo/priv/lib"
```

Comment:
Are these bin and lib saved across different runs (e.g. distinct commits) or across different stages of the same run (e.g. build then test)? My concern that change in build env across commits may make these binaries cause failure.

Reply:
Ah, I tried to use workspaces for this at first, then after switching to cache I forgot to use a local scope key. Fixed in 33e2fba.